### ISSUE TITLE
Add support for /announce[color] command variations

### ISSUE THREAD

**Issue by Mm2PL** [Chatterino/chatterino2#5218]
### Discussed in https://github.com/Chatterino/chatterino2/discussions/4367

<div type='discussions-op-text'>

<sup>Originally posted by **Pafon22** February 11, 2023</sup>
Unrecognized command when sending /announce with any color.
Insert /announceblue /announcegreen /announceorange /announcepurple in the options when possible. I don't know if Twitch already has the endpoint for this.</div>

However first colored announcement support needs to be implemented. I can see this functionality being merged into a single PR if neither part is too large.

**Comment by pajlada** [comment]
When implementing the commands, they should replicate Twitch chat's behaviour
![image](https://github.com/Chatterino/chatterino2/assets/962989/7aec2c6a-9cfe-4b36-90ce-d5ef46c5b2af)

**Comment by 4rneee** [comment]
Supporting the `/announce[color]` commands would be very straight forward and I could create a PR for that.  
I also looked a bit into displaying colored announcements and don't know how (or if) this can be implemented because I didn't find any documentation. https://dev.twitch.tv/docs/irc/tags/#usernotice-tags doesn't even document the `announcement` notice in the `msg-id` tag which we currently use to detect announcements.

**Comment by pajlada** [comment]
I'm ok with implementing sending, even if receiving colored announcements don't work

**Comment by Mm2PL** [comment]
Here are some example announcement messages:

<details>
<summary>Examples</summary>

```
@badge-info=;badges=broadcaster/1,no_audio/1;color=#DAA520;display-name=Mm2PL;emotes=;flags=;id=98928e18-2c4c-4914-97ea-032a8b47d94c;login=mm2pl;mod=0;msg-id=announcement;msg-param-color=BLUE;room-id=117691339;subscriber=0;system-msg=;tmi-sent-ts=1710346144413;user-id=117691339;user-type=;vip=0 :tmi.twitch.tv USERNOTICE #mm2pl :blue
@badge-info=;badges=broadcaster/1,no_audio/1;color=#DAA520;display-name=Mm2PL;emotes=;flags=;id=c3e34a2e-b7be-4608-a8b2-18c6c455c302;login=mm2pl;mod=0;msg-id=announcement;msg-param-color=GREEN;room-id=117691339;subscriber=0;system-msg=;tmi-sent-ts=1710346157120;user-id=117691339;user-type=;vip=0 :tmi.twitch.tv USERNOTICE #mm2pl :green
@badge-info=;badges=broadcaster/1,no_audio/1;color=#DAA520;display-name=Mm2PL;emotes=;flags=;id=dcf6e51b-5e11-4f34-8aef-8a27dc890592;login=mm2pl;mod=0;msg-id=announcement;msg-param-color=ORANGE;room-id=117691339;subscriber=0;system-msg=;tmi-sent-ts=1710346160360;user-id=117691339;user-type=;vip=0 :tmi.twitch.tv USERNOTICE #mm2pl :orange
@badge-info=;badges=broadcaster/1,no_audio/1;color=#DAA520;display-name=Mm2PL;emotes=;flags=;id=646a9777-9e23-4e4a-85c9-171a56766e0d;login=mm2pl;mod=0;msg-id=announcement;msg-param-color=PURPLE;room-id=117691339;subscriber=0;system-msg=;tmi-sent-ts=1710346162937;user-id=117691339;user-type=;vip=0 :tmi.twitch.tv USERNOTICE #mm2pl :purple
@badge-info=;badges=broadcaster/1,no_audio/1;color=#DAA520;display-name=Mm2PL;emotes=;flags=;id=d93a10ba-d79e-4177-ae72-35ffddbf4833;login=mm2pl;mod=0;msg-id=announcement;msg-param-color=PRIMARY;room-id=117691339;subscriber=0;system-msg=;tmi-sent-ts=1710346166508;user-id=117691339;user-type=;vip=0 :tmi.twitch.tv USERNOTICE #mm2pl :my primary color #DAA520
```

</details>


The tag for the color is  `msg-param-color`.

> system-msg tag is always empty, the user text is only in the message content, 
> msg-param-color IS NOT a hex color, don't be fooled, it's an enum of `BLUE`, `GREEN`, `ORANGE`, `PURPLE`, `PRIMARY` (this value means to use the `color` tag)

**Comment by 4rneee** [comment]
Okay, thanks for the info. I opened a PR which only implements sending the commands.

Using your information I will further look into displaying colored announce messages and open another PR once I implemented it.

**Comment by 4rneee** [comment]
I thought of taking a look at this again in the near future. @pajlada does this make sense before the highlight rework or should I wait for that to be merged?

**Comment by pajlada** [comment]
> I thought of taking a look at this again in the near future. [@pajlada](https://github.com/pajlada) does this make sense before the highlight rework or should I wait for that to be merged?

Feel free to take a look at this, it should only somewhat touch highlights and I can deal with some merge conflicts

**Comment by 4rneee** [comment]
So, I came up with two possible approaches:

1. I only add one message flag for generic "Announcement" and store the color of the announcement elsewhere (maybe as enum in `Message`?). Then the user can configure a color for announcements. Additionally I would add a setting (not sure where exactly) to configure whether or not to show "colored announcements". If this is on we use default colors for e.g. `/announceblue` and the configured color for "Primary"(which is when just `/announce` is used).
2. Alternatively I could create a messageflag for each color and let the user configure every color.

I currently tend to the first one since it does not waste that many bits in `MessageFlag` and it feels weird to be able to configure the color of a "blue announcement".

Any thoughts or additional ideas?

**Comment by Nerixyz** [comment]
> I currently tend to the first one since it does not waste that many bits in `MessageFlag` and it feels weird to be able to configure the color of a "blue announcement".

+1. Also, if you'd have a bit for every color, you could have weird multicolor announcements.

The announcement color field is one of these "extra" fields in a `Message`. I'd love to have these fields packed in some variant or extra data/union. Currently, you can have a stacked timeout system message with bits and a reward.

**Comment by pajlada** [comment]
yeah option 1 is reasonable. no need to think about the variant/extra data stuff - just plonk the announcement color in Message for now
